### PR TITLE
[fix][broker] Fix broker cache eviction of entries read by active cursors

### LINKED ISSUE #9958: [Performance] Optimize allocations in ManagedLedgerImpl.getEarlierReadPositionForActiveCursors()/ManagedCursorContainer.iterator()
### Problem

In profiling, I can see that a lot of allocations in TLAB are made for this kind of stack trace:
```
Iterator java.util.concurrent.ConcurrentSkipListMap.entryIterator()
Iterator java.util.concurrent.ConcurrentSkipListMap$EntrySet.iterator()
Iterator org.apache.bookkeeper.mledger.impl.ManagedCursorContainer.iterator()
PositionImpl org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl.getEarlierReadPositionForActiveCursors()
void org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl.doCacheEviction(long)
void org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl.lambda$doCacheEviction$5(long
void org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl$$Lambda$83.1547081970.accept(Object)
void java.util.concurrent.ConcurrentHashMap$ValuesView.forEach(Consumer)
void org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl.doCacheEviction()
void org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl.cacheEvictionTask()
void org.apache.bookkeeper.mledger.impl.ManagedLedgerFactoryImpl$$Lambda$81.1649498695.run()
void java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor$Worker)
void java.util.concurrent.ThreadPoolExecutor$Worker.run()
void io.netty.util.concurrent.FastThreadLocalRunnable.run()
void java.lang.Thread.run()
```
(Broker version is Pulsar 2.7.0)

### Expected behavior

Calls on the hot paths shouldn't create a lot of allocations.

### LINKED ISSUE #16054: Broker cache eviction evicts entries that haven't been read by active consumers
**Describe the bug**

There's a performance regression with large fan-outs (large number of consumers).

#12045 changed cache eviction. The problem is that cache eviction is done to the position whatever happens to be the first cursor in the "heap" list in ManagedCursorContainer. There's a chance for a race condition: another cursor in the list might have an older position, but that hasn't been updated in the ordered data structure ("heap") when the method is called. The cache eviction will take the read position directly from the first cursor and it ignores and older positions.
When the consumer for the other cursor does a read, the entries are no more in the cache and there will be a cache miss. 

**To Reproduce**

It should be possible to reproduce this in a multi-consumer scenario where the speed of the consumers varies.

**Expected behavior**

Cache eviction should properly consider all active consumers. This was the state before #12045 and #14985 changes.

**Additional context**
#12045, #14985, #16049

This impacts all maintenance branches since #12045 was cherry-picked to branch-2.7 (2.7.4), branch-2.8 (2.8.2) and branch-2.9 (2.9.0) at the time.

### PR DESCRIPTION
**Pull request by lhotari**
Fixes #16054
Fixes #9958

### Motivation

The broker cache eviction of entries read by active cursors has been broken in Pulsar since 2.8.2 version.
This broke in the PR #12045 changes. It was a change to optimize the broker cache behavior where there was a high over head of the eviction task (issue reported as #9958 and also reported on Pulsar Slack by multiple users during that time).
PR #12045 change was partially mitigated by a new feature "cacheEvictionByMarkDeletedPosition" added by PR #14985. However that introduces an issue that entries could get cached for too long and unnecessarily. 

The main motivation of this PR is to fix and restore broker cache eviction of entries based on the earliest read position of active cursors (consumers) and also handle this in an efficient and performant way so that the original intention of PR #12045 will be covered, the performance issue reported as #9958.

### Modifications

- add a new ManagedCursorContainer instance to ManagedLedgerImpl that keeps a sorted list of cursors ordered by the read position
- make changes to ManagedLedgerImpl and ManagedCursorImpl to notify the ledger instance when the cursor's read position gets updated. This makes it possible to react to the changes and sort the cursors only when there has been changes to the data that is used for sorting. This reactive approach has been used for the mark delete position and this exists. It fixes the performance issue in the most efficient way.

- There were some hard to understand tests added in #14985. Those tests have been replaced by simpler tests that will prevent future regressions.
- Some tests contained invalid assertions perhaps caused by earlier bugs in the eviction by read position. This has been fixed.
- More tests have been added to cover possible corner cases

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 1]
```diff
@@ -166,8 +166,9 @@ public class ManagedLedgerImpl implements ManagedLedger, CreateCallback {
 
     private final ManagedCursorContainer cursors = new ManagedCursorContainer();
     private final ManagedCursorContainer activeCursors = new ManagedCursorContainer();
```
Comment (eolivelli) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:168]: what about adding a comment here and say that this is ordered by the delete position ?

**Review by eolivelli — APPROVED** [review]
LGTM

great work

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 1 above)
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:170]: yes, makes sense. added some comments

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 2]
```diff
@@ -2182,21 +2187,12 @@ public boolean hasMoreEntries(PositionImpl position) {
...
-            entryCache.invalidateEntries(pair.getRight());
-        }
-    }
-
     public PositionImpl getEvictionPosition(){
         PositionImpl evictionPos;
         if (config.isCacheEvictionByMarkDeletedPosition()) {
-            PositionImpl earlierMarkDeletedPosition = getEarlierMarkDeletedPositionForActiveCursors();
-            evictionPos = earlierMarkDeletedPosition != null ? earlierMarkDeletedPosition.getNext() : null;
+            evictionPos = activeCursors.getSlowestReaderPosition();
```
Comment (nodece) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2193]: Miss offset?

Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2193]: @nodece activeCursors contains the active cursors ordered by mark delete position. What did you mean with your question?

Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2193]: the tests in ManagedLedgerTest also verify the behavior

Comment (nodece) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2193]: Old behavior is `earlierMarkDeletedPosition.getNext()`, new behavior without `getNext()`.

Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2193]: the tests confirm that it's correct now. I believe that it was invalid previously.

**Comment by lhotari** [comment]
I marked this as draft until I have fixed the non-durable cursor behavior. The intended behavior is described in #6787 where this was added. 

It seems that broker caching should take non-durable cursors into account, but it shouldn't be taken into account for trimming ledgers. 

#12045 fixed the behavior for non-durable cursor caching as a side-effect. However, that wasn't explained in the PR itself.

**Comment by lhotari** [comment]
This is ready for review now.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 3]
```diff
@@ -3513,31 +3495,22 @@ public void activateCursor(ManagedCursor cursor) {
...
-            if (activeCursors.get(cursor.getName()) != null) {
-                activeCursors.removeCursor(cursor.getName());
-                if (!activeCursors.hasDurableCursors()) {
-                    // cleanup cache if there is no active subscription
-                    entryCache.clear();
-                } else {
-                    // if removed subscription was the slowest subscription : update cursor and let it clear cache:
-                    // till new slowest-cursor's read-position
-                    discardEntriesFromCache((ManagedCursorImpl) activeCursors.getSlowestReader(),
-                            getPreviousPosition((PositionImpl) activeCursors.getSlowestReader().getReadPosition()));
```
Comment (lhotari) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:3532]: @michaeljmarshall @cdbartholomew These are the lines that I think that were completely invalid previously and could lead to the cache being evicted unintentionally.

**Review by eolivelli — APPROVED** [review]

**Comment by lhotari** [comment]
rebased over #17300 to get builds going before it's merged.

**Review by codelipenghui — APPROVED** [review]
LGTM, left some minor comments.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 4]
```diff
@@ -2182,63 +2180,33 @@ public boolean hasMoreEntries(PositionImpl position) {
         return result;
     }
 
-    void discardEntriesFromCache(ManagedCursorImpl cursor, PositionImpl newPosition) {
-        Pair<PositionImpl, PositionImpl> pair = activeCursors.cursorUpdated(cursor, newPosition);
-        if (pair != null) {
-            entryCache.invalidateEntries(pair.getRight());
+    void doCacheEviction(long maxTimestamp) {
+        if (entryCache.getSize() > 0) {
+            entryCache.invalidateEntriesBeforeTimestamp(maxTimestamp);
```
[the remark this answers is not in the document]
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2185]: It is no more necessary to call it in `doCacheEviction` method since `invalidateEntriesUpToSlowestReaderPosition` will get called when the slowest reader position changes.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 5]
```diff
@@ -2182,63 +2180,33 @@ public boolean hasMoreEntries(PositionImpl position) {
...
+            entryCache.clear();
         }
-        return durablePosition.compareTo(nonDurablePosition) > 0 ? nonDurablePosition : durablePosition;
     }
 
     void updateCursor(ManagedCursorImpl cursor, PositionImpl newPosition) {
         Pair<PositionImpl, PositionImpl> pair = cursors.cursorUpdated(cursor, newPosition);
+        if (config.isCacheEvictionByMarkDeletedPosition()) {
+            updateActiveCursor(cursor, newPosition);
+        }
```
[the remark this answers is not in the document]
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2209]: makes sense. I'll rename the methods.

[the remark this answers is not in the document]
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: done.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (codelipenghui) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:2185]: Oh, got it. Thanks.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 6]
```diff
@@ -3513,31 +3495,22 @@ public void activateCursor(ManagedCursor cursor) {
...
-                    // if removed subscription was the slowest subscription : update cursor and let it clear cache:
-                    // till new slowest-cursor's read-position
-                    discardEntriesFromCache((ManagedCursorImpl) activeCursors.getSlowestReader(),
-                            getPreviousPosition((PositionImpl) activeCursors.getSlowestReader().getReadPosition()));
-                }
-            }
-            if (!cursor.isDurable()) {
-                nonDurableActiveCursors.removeCursor(cursor.getName());
+            if (activeCursors.get(cursorName) != null) {
+                activeCursors.removeCursor(cursorName);
```
Comment (michaeljmarshall) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:3507]: Nit: could we update the `removeCursor` to could return a `boolean` based on whether the cursor was removed? Then we'd save the `get` call on 3507 which internally requires us to acquire the `rwLock`.

**Review by michaeljmarshall — APPROVED** [review]
LGTM.

I think we should update the javadoc for `ManagedCursorContainer`. Also, we could probably make a minor optimization in the `deactivateCursorByName` method.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (michaeljmarshall) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: Great suggestion @codelipenghui.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 6 above)
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: Thanks for the suggestion. I made this change.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorContainer.java [hunk 7]
```diff
@@ -45,33 +45,48 @@
...
     }
 
-    public enum CursorType {
-        DurableCursor,
-        NonDurableCursor,
-        ALL
+    private enum OrderingType {
+        MARKDELETED_POSITION,
+        READ_POSITION
+    }
```
[the remark this answers is not in the document]
Comment (lhotari) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorContainer.java]: I updated the Javadoc and ended up simplifying ManagedCursorContainer. I removed that OrderingType and OrderingFilter completely and instead modified the add method to explicitly pass the position which is used for ordering. If this is null, the cursor won't be tracked in the heap and ordered. This feels more consistent with the "cursorUpdated" method which also takes a cursor and position parameters.

**Review by michaeljmarshall — APPROVED** [review]
LGTM. I like the updated design @lhotari.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorContainer.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (michaeljmarshall) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorContainer.java]: Makes sense to me. I like the update.

**Comment by mattisonchao** [comment]
Hello @lhotari 
Could you help to cherry-pick this PR to branch-2.9 (To avoid cherry-picking involving bugs)

**Comment by HQebupt** [comment]
> Hello @lhotari
> Could you help to cherry-pick this PR to branch-2.9 (To avoid cherry-picking involving bugs)

I can do it if you agree. @lhotari

**Comment by congbobo184** [comment]
could you please cherry-pick this PR to branch-2.9? thanks.
